### PR TITLE
Tests and changes for CRAM support.

### PR DESCRIPTION
**Pull request by cmnbroad**
There is one test in ValidateSamFileIntegrationTest that is commented out since it depends on a change to htsjdk.

### CODE REVIEW

**Comment by droazen** [comment]
@cmnbroad told me that this one isn't quite ready for review yet, so we should wait to review until he gives the word

**Comment by akiezun** [comment]
assigning to @cmnbroad then until it's ready for review

**Comment by droazen** [comment]
@cmnbroad pinging you on this one -- is this ready for review?

**Comment by cmnbroad** [comment]
This is ready for review. I'm creating/updating tickets for the remaining tools that have issues. There are also a couple of cases where there are tests commented out because they need fixes in htsjdk (GatherBamFiles and SamFileValidator at least - the fixes are all in PR https://github.com/samtools/htsjdk/pull/368) and I'll make tickets to reflect those as well.

**Comment by droazen** [comment]
@akiezun can you review?

File: src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java [hunk 1]
```diff
@@ -67,26 +68,96 @@ public String getTestedClassName() {
...
+        return new Object[][]{
+                // these require the fix to https://github.com/samtools/htsjdk/issues/365
+                // in PR https://github.com/samtools/htsjdk/pull/368
+                //{splitCramsUnmappedLast, origCram, true, true},
+                //{splitCramsUnmappedFirst, origCram, true, true},
+                //{splitCramsUnmappedMissing, origCram, false, false}
+        };
+    }
+
+    // These tests require the fix to https://github.com/samtools/htsjdk/issues/365
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/sam/GatherBamFilesIntegrationTest.java]: I would prefer to have those disabled by `enabled=false`. Please enter a ticket for gatk as well - to remember to reenable them

File: src/test/java/org/broadinstitute/hellbender/tools/picard/sam/SortSamIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+public class SortSamIntegrationTest extends CommandLineProgramTest {
+
+    @DataProvider(name="sortbams")
+    public Object[][] sortBAMData() {
+        return new Object[][] {
+                {"count_reads.bam", "count_reads_sorted.bam", null, ".bam", "coordinate"},
+                {"count_reads.bam", "count_reads_sorted.bam", "count_reads.fasta", ".cram", "coordinate"},
+                {"count_reads.bam", "count_reads.bam", null, ".bam", "queryname"},
+                // next two require fix to https://github.com/samtools/htsjdk/issues/365
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/sam/SortSamIntegrationTest.java]: here are 2 more disabled tests that need to be mentioned in that ticket (see above)

File: src/test/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFileIntegrationTest.java [hunk 3]
```diff
@@ -20,22 +21,34 @@ public String getTestedClassName() {
...
-            {"valid.sam", "VERBOSE", true},
-            {"invalid_coord_sort_order.sam", "SUMMARY", false},
+          {"valid.sam", "SUMMARY", null, true},
+          {"valid.sam", "VERBOSE", null, true},
+          {"valid.bam", "VERBOSE", null, true},
+          {"valid.bam", "VERBOSE", "valid.fasta", false}, // the NM tags in this sam file don't match the reference deltas
+          {"valid.cram", "VERBOSE", "valid.fasta", true}, // the NM tags in the CRAM file do match the reference (ty samtools)
+          {"invalid_coord_sort_order.sam", "SUMMARY", null, false},
+          {"invalid_coord_sort_order.sam", "SUMMARY", "invalid_coord_sort_order.fasta", false},
+          //{"invalid_coord_sort_order.cram", "SUMMARY", "invalid_coord_sort_order.fasta", false},  // requires htsjdk fix to SamFileValidator
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/picard/sam/ValidateSamFileIntegrationTest.java]: We need a ticket in gatk to reenable this (that ticket may refer to the relevent issue in htsjdk)

**Comment by akiezun** [comment]
this looks good to me :+1: But please enter tickets to reenable the commented out tests and rebase/squash/merge when ready.

**Comment by cmnbroad** [comment]
There are already tickets for two of these (#1138 and #1141); I updated #1141 to include the SortSam tests so these are all covered.

**Pull request merged**